substrate/client/network/src/litep2p/shim/notification/peerset.rs

```diff
@@ -0,0 +1,1276 @@
+			// ([`PeerState::Closing`]) and it was a non-reserved peer
+			PeerState::Connected { direction: Direction::Outbound(Reserved::No) } |
+			PeerState::Closing { direction: Direction::Outbound(Reserved::No) } => {
+				log::trace!(
+					target: LOG_TARGET,
+					"{}: outbound substream closed to non-reserved peer {peer:?} {state:?}",
+					self.protocol,
+				);
+
+				adjust_or_warn!(
+					self.num_out,
+					peer,
+					self.protocol,
+					Direction::Outbound(Reserved::No)
+				);
+			},
+			// reserved peers don't require adjustments to slot counts
+			PeerState::Closing { .. } | PeerState::Connected { .. } => {
+				log::debug!(target: LOG_TARGET, "{}: reserved peer {peer:?} disconnected", self.protocol);
+			},
+			state => {
+				log::warn!(target: LOG_TARGET, "{}: invalid state for disconnected peer {peer:?}: {state:?}", self.protocol);
+				debug_assert!(false);
+			},
+		}
+		*state = PeerState::Backoff;
+
+		self.connected_peers.fetch_sub(1usize, Ordering::Relaxed);
+		self.pending_backoffs.push(Box::pin(async move {
+			Delay::new(DEFAULT_BACKOFF).await;
+			(peer, DISCONNECT_ADJUSTMENT)
+		}));
+	}
+
+	/// Report to [`Peerset`] that an inbound substream was opened and that it should validate it.
+	pub fn report_inbound_substream(&mut self, peer: PeerId) -> ValidationResult {
+		log::trace!(target: LOG_TARGET, "{}: inbound substream from {peer:?}", self.protocol);
+
+		let state = self.peers.entry(peer).or_insert(PeerState::Disconnected);
+		let reserved_peer = self.reserved_peers.contains(&peer);
+
+		match state {
+			// disconnected peers proceed directly to inbound slot allocation
+			PeerState::Disconnected => {},
+			// peer is backed off but if it can be accepted (either a reserved peer or inbound slot
+			// available), accept the peer and then just ignore the back-off timer when it expires
+			PeerState::Backoff => {
+				if !self.reserved_peers.contains(&peer) && self.num_in == self.max_in {
+					log::trace!(
+						target: LOG_TARGET,
+						"{}: ({peer:?}) is backed-off and cannot accept, reject inbound substream",
+						self.protocol,
+					);
+
+					return ValidationResult::Reject
+				}
+			},
+			// `Peerset` had initiated an outbound substream but litep2p had received an inbound
+			// substream before the command was received. As litep2p is the source of truth as far
+			// as substream states go, `Peerset` must update its own state and attempt to accept the
+			// peer as inbound if there are enough slots.
+			PeerState::Opening { direction: Direction::Outbound(reserved) } => {
+				log::trace!(
+					target: LOG_TARGET,
+					"{}: inbound substream received for {peer:?} that was marked outbound",
+					self.protocol,
+				);
+
+				match (reserved_peer, &reserved) {
+					(true, &Reserved::Yes) => {},
+					(false, &Reserved::No) => {},
+					_ => panic!(
+						"{}: state mismatch for {peer:?}, {reserved_peer} {reserved:?}",
+						self.protocol
```

Comment:
Can't this backfire if the set of reserved peers was modified after the peer was put into `Opening` state?

Reply:
It shouldn't backfire because the `reserved_peers` and the peer state are updated together. But I think `panic!()` is too harsh so I'll downgrade it to a warning + `debug_assert!()`